pkg/controller/elasticsearch/user/filerealm/role_users.go

```diff
@@ -0,0 +1,85 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package filerealm
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/elastic/cloud-on-k8s/pkg/utils/set"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/stringsutil"
+)
+
+// roleUsersMapping is a map {role name -> [] user names}
+type roleUsersMapping map[string][]string
+
+// mergeWith merges multiple usersPasswordHashes, giving priority to other.
+func (r roleUsersMapping) mergeWith(other roleUsersMapping) roleUsersMapping {
+	if len(r) == 0 {
+		return other
+	}
+	if len(other) == 0 {
+		return r
+	}
+	for otherRole, otherUsers := range other {
+		currentUsers, exists := r[otherRole]
+		if !exists {
+			// role does not exist yet, create it
+			r[otherRole] = otherUsers
+			continue
+		}
+		// role already exists, merge sorted users
+		userSet := set.Make(currentUsers...)
+		userSet.MergeWith(set.Make(otherUsers...))
+		userSlice := userSet.AsSlice()
+		stringsutil.SortStringSlice(userSlice)
+		r[otherRole] = userSlice
+	}
+	return r
+}
+
+// fileBytes serializes the roleUsersMapping into a file with format:
+// ```
+// role1:user1,user2,user3
+// role2:user1
+// ```
+// Rows are sorted for easier comparison.
+func (r roleUsersMapping) fileBytes() []byte {
+	rows := make([]string, 0, len(r))
+	for role, users := range r {
+		stringsutil.SortStringSlice(users)
```

Comment:
Just for my understanding: you are sorting here again in case the `roleUsersMapping` was manually constructed and not via `mergeWith` which already sorts the users slices?

Reply:
Hmm you're right, sorting is not very consistent in this file.
I think the following would make sense:
- the internal representation should always contain sorted users
- when we parse the raw file, we sort users
- when we merge 2 `roleUsersMapping`, we should expect both already have sorted users
- when we serialize back to bytes, we don't have to re-sort users since they're already sorted

I'll change the code accordingly.